### PR TITLE
Updated SoftmaxCrossEntropy interface & cleaned Softmax, Cross Entropy layers

### LINKED ISSUE #12: Feature Requested: Change SoftmaxCrossEntropyLoss layer to accept Float matrix label

### PR DESCRIPTION
**Pull request by fufjvnvnf**
Resolves #12

### CODE REVIEW

**Comment by fufjvnvnf** [comment]
This PR, namely the changes on the interface is likely to break `mnist_conv_softmax.jl` and `mnist_cnn.jl`, but since I'm next going to change the prediction interface of the loss functions I'll fix those main classes in the next PR.

**Comment by stevenygd** [comment]
num_iter = 500
t_, b_ = benchmmark(smaxcross, smax, cross, 256, 1024)

Before:
Old forward uses on average 0.00635987514 seconds
New forward uses on average 0.010269467555999997 seconds
Old backward uses on average 0.005357705962000001 seconds
New backward uses on average 0.013618935071999986 seconds
Old forward allocates on average 4.245799296e6 bytes
New forward allocates on average 8.3952e6 bytes
Old backward allocates on average 2.103335296e6 bytes
New backward allocates on average 8.3952e6 bytes

After:
Old forward uses on average 0.008961172802000013 seconds
New forward uses on average 0.012715808534000003 seconds
Old backward uses on average 0.0010959743539999994 seconds
New backward uses on average 0.008850074861999997 seconds
Old forward allocates on average 8.397008e6 bytes
New forward allocates on average 8.3952e6 bytes
Old backward allocates on average 4.197328e6 bytes
New backward allocates on average 8.3952e6 bytes

**Comment by fufjvnvnf** [comment]
Actual result in the same size, in Julia 6.
Before:
```
Old forward uses on average 0.030599784640000002 seconds
New forward uses on average 0.013083654519999996 seconds
Old backward uses on average 0.008814251459999999 seconds
New backward uses on average 0.013510984159999996 seconds
Old forward allocates on average 4.26784896e6 bytes
New forward allocates on average 8.417728e6 bytes
Old backward allocates on average 2.11386496e6 bytes
New backward allocates on average 8.417728e6 bytes
```
After:
```
Old forward uses on average 0.011977900916000003 seconds
New forward uses on average 0.014186884568000008 seconds
Old backward uses on average 0.001087744465999998 seconds
New backward uses on average 0.009409748012000006 seconds
Old forward allocates on average 8.421442596e6 bytes
New forward allocates on average 8.418128e6 bytes
Old backward allocates on average 4.197616e6 bytes
New backward allocates on average 8.418128e6 bytes
```

**Comment by stevenygd** [comment]
@fufjvnvnf It occurs to me that there are some conflicts. Please try to resolve it first. Thx!

**Comment by fufjvnvnf** [comment]
@stevenygd Looks like the conflicts have been resolved. It also occurs to me that loop does not help in improving the speed for both forward and backward. Could this branch be merged for now since it is faster than the one on the master branch, and we would look back on further improving the speed in the future?

**Comment by stevenygd** [comment]
@fufjvnvnf I'm a little bit reluctant on that, since slow layers will eventually bite us back at the end. But just for this case. Please add some more comments on your approaches, and fix the conflict. After you've done that, I will merge this PR.

**Comment by fufjvnvnf** [comment]
Is there still conflict on this pr? I see the 'no-conflict' sign on my view.
